services/worker/src/worker/job_runners/split/descriptive_stats.py

```diff
@@ -0,0 +1,288 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2023 The HuggingFace Authors.
+
+import logging
+from pathlib import Path
+from typing import Dict, List, Literal, Optional, Tuple, TypedDict, Union
+
+import duckdb
+import numpy as np
+from libcommon.constants import PROCESSING_STEP_SPLIT_DESCRIPTIVE_STATS_VERSION
+from libcommon.exceptions import PreviousStepFormatError
+
+# from libcommon.exceptions import SplitWithTooBigParquetError
+from libcommon.processing_graph import ProcessingStep
+from libcommon.simple_cache import get_previous_step_or_raise
+from libcommon.utils import JobInfo
+from tqdm import tqdm
+
+from worker.config import AppConfig
+from worker.job_runners.split.split_job_runner import SplitCachedJobRunner
+from worker.utils import CompleteJobResult
+
+DECIMALS = 5
+
+INTEGER_DTYPES = ["int8", "int16", "int32", "int64"]
+FLOAT_DTYPES = ["float16", "float32", "float64"]
+
+
+SplitDescriptiveStatsJobRunnerErrorCode = Literal["PreviousStepFormatError"]
+
+
+class Histogram(TypedDict):
+    hist: List[int]
+    bin_edges: List[float]
+
+
+class NumericalStatsItem(TypedDict):
+    nan_count: int
+    nan_prop: float
+    min: float
+    max: float
+    mean: float
+    median: float
+    std: float
+    histogram: Histogram
+
+
+class CategoricalStatsItem(TypedDict):
+    nan_count: int
+    nan_prop: float
+    n_unique: int
+    frequencies: Dict[str, int]
+
+
+class StatsPerColumnItem(TypedDict):
+    column_name: str
+    column_type: str
+    column_dtype: Optional[str]
+    column_stats: Union[NumericalStatsItem, CategoricalStatsItem]
+
+
+class SplitDescriptiveStatsResponse(TypedDict):
+    num_examples: int
+    stats: List[StatsPerColumnItem]
+
+
+def compute_histogram(
+    con: duckdb.DuckDBPyConnection,
+    column_name: str,
+    urls: List[str],
+    bin_size: int,
+) -> Histogram:
+    hist_query = f"""
+    SELECT FLOOR("{column_name}"/{bin_size})*{bin_size}, COUNT(*) FILTER (WHERE {column_name} IS NOT NULL)
+     FROM read_parquet({urls}) GROUP BY 1 ORDER BY 1;
```

Comment:
Not sure if it would help, but I am seeing that everywhere it is reading the parquets from HTTP, would it be helpful if first we copy the info in a local db? In the future, this step and `split-duckdb-index` could depend on another one where first: all data is ingested in a duckdb db file and then, everything is run against duckdb engine directly. 
Just a random idea (feel free to ignore it)